### PR TITLE
Kiwi Editor - project repositories

### PR DESCRIPTION
**Pull request by DavidKang**
We had a checkbox where the users can check it to use the repositories from the current project.

screenshot from 2017-09-05 13-04-46

When this option is checked we hide the others repositories because it's not allowed to use `obsrepositories` with others repositories at the same time. When the whole Kiwi Image is saved, these repositories will be deleted.

screenshot from 2017-09-05 15-31-19

screenshot from 2017-09-06 12-40-34

Also when we try to import a kiwi file with multiple repositories and one of them has as `source_path='obsrepositories:/'`, it will be redirected to the view file to indicated what happen.

screenshot from 2017-09-06 15-04-11

### CODE REVIEW

File: src/api/app/controllers/webui/kiwi/images_controller.rb [hunk 1]
```diff
@@ -87,6 +90,13 @@ def set_image
       def authorize_update
         authorize @image, :update?
       end
+
+      def cleanup_non_project_repositories!
+        return unless params[:kiwi_image][:use_project_repositories] == '1'
+
+        @image.repositories.delete_all
```
Comment (ChrisBr) [inline src/api/app/controllers/webui/kiwi/images_controller.rb]: Just as a thought which we can maybe implement later:
What about not deleting the repositories but instead marking them as ``disabled`` and when the user disables ``obsrepositories`` we add these repositories again?

File: src/api/app/models/kiwi/image.rb [hunk 2]
```diff
@@ -48,43 +49,51 @@ def self.build_from_xml(xml_string, md5)
...
-      attributes['imageinclude'] = repository['imageinclude'] == 'true' if repository.key?('imageinclude')
-      attributes['prefer_license'] = repository['prefer-license'] == 'true' if repository.key?('prefer-license')
-
-      new_image.repositories.build(attributes)
-      order += 1
+
+    if repositories
+      new_image.use_project_repositories = true if repositories.map { |repository| repository['source']['path'] }.include?('obsrepositories:/')
+
+      repositories.reject{ |repository| repository['source']['path'] == 'obsrepositories:/' }.each do |repository|
```
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: ``repositories.reject{ |repository| repository['source']['path'] == 'obsrepositories:/' }.each.with_index(1) do |repository, index|``

and you can get rid of the order variable.

File: src/api/app/models/kiwi/image.rb [hunk 3]
```diff
@@ -48,43 +49,51 @@ def self.build_from_xml(xml_string, md5)
...
     package_groups = [xml["packages"]] if xml["packages"].is_a?(Hash)
-    package_groups.each do |package_group_xml|
-      attributes = {
-        kiwi_type:    package_group_xml['type'],
-        profiles:     package_group_xml['profiles '],
-        pattern_type: package_group_xml['patternType']
-      }
-      package_group = Kiwi::PackageGroup.new(attributes)
-      package_group_xml['package'].each do |package|
+    if package_groups
```
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: Why is this necessary now? I think it should always be an array, no?

File: src/api/app/models/kiwi/image.rb [hunk 4]
```diff
@@ -48,43 +49,51 @@ def self.build_from_xml(xml_string, md5)
...
-        password:    repository['password']
-      }
-      attributes['imageinclude'] = repository['imageinclude'] == 'true' if repository.key?('imageinclude')
-      attributes['prefer_license'] = repository['prefer-license'] == 'true' if repository.key?('prefer-license')
-
-      new_image.repositories.build(attributes)
-      order += 1
+
+    if repositories
+      new_image.use_project_repositories = true if repositories.map { |repository| repository['source']['path'] }.include?('obsrepositories:/')
```
Comment (ChrisBr) [inline src/api/app/models/kiwi/image.rb]: ``
new_image.use_project_repositories = repositories.any? { |repository| repository['source']['path'] == 'obsrepositories:/'
``

Comment (ChrisBr) [reply src/api/app/models/kiwi/image.rb]: Btw: Have you tried if ``obsrepositories://`` is also valid?

File: src/api/spec/controllers/webui/kiwi/images_controller_spec.rb [hunk 5]
```diff
@@ -47,9 +48,48 @@
...
+          let(:package_with_kiwi_file) do
+            create(:package_with_kiwi_file, name: 'package_with_invalid_kiwi_file',
+                   project: project, kiwi_file_content: invalid_kiwi_xml_with_obsrepositories)
+          end
+
+          before do
+            get :import_from_package, params: { package_id: package_with_kiwi_file.id }
+          end
+
+          it 'redirect to package_view_file_path' do
```
Comment (ChrisBr) [inline src/api/spec/controllers/webui/kiwi/images_controller_spec.rb]: What about checking the flash?

File: src/api/app/models/kiwi/image.rb (hunk 4 above)
Comment (DavidKang) [reply src/api/app/models/kiwi/image.rb]: @ChrisBr, `obsrepositories://` is not valid, should it be valid?

**Comment by adrianschroeter** [comment]
On Montag, 11. September 2017, 13:42:44 CEST wrote David Kang:
> DavidKang commented on this pull request.
> 
> 
> 
> > -        source_path: repository['source']['path'],
> -        priority:    repository['priority'],
> -        order:       order,
> -        alias:       repository['alias'],
> -        replaceable: repository['status'] == 'replaceable',
> -        username:    repository['username'],
> -        password:    repository['password']
> -      }
> -      attributes['imageinclude'] = repository['imageinclude'] == 'true' if repository.key?('imageinclude')
> -      attributes['prefer_license'] = repository['prefer-license'] == 'true' if repository.key?('prefer-license')
> -
> -      new_image.repositories.build(attributes)
> -      order += 1
> +
> +    if repositories
> +      new_image.use_project_repositories = true if repositories.map { |repository| repository['source']['path'] }.include?('obsrepositories:/')
> 
> @ChrisBr, `obsrepositories://` is not valid, should it be valid?

no, it shouldn't.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

File: src/api/app/models/kiwi/image.rb (hunk 3 above)
Comment (DavidKang) [reply src/api/app/models/kiwi/image.rb]: @ChrisBr, It was always necessary because `package_groups = xml["packages"]` return `nil` if there not exist any packages. The same happens with repositories.

File: src/api/app/models/kiwi/image.rb (hunk 4 above)
Comment (ChrisBr) [reply src/api/app/models/kiwi/image.rb]: As @adrianschroeter wrote it shouldn't be valid, everything fine!

File: src/api/app/controllers/webui/kiwi/images_controller.rb (hunk 1 above)
Comment (DavidKang) [reply src/api/app/controllers/webui/kiwi/images_controller.rb]: @ChrisBr, It's a good idea, I will create a new card to do this in another PR.

**Comment by ChrisBr** [comment]
Added the "DO NOT MERGE" label as this PR contains a migration and we agreed that we only run migrations Thursday morning in the maintenance window. Fell free to remove and merge this Wednesday evening to have it ready for deployment on Thursday morning.

**Comment by ChrisBr** [comment]
LGTM :+1: :heart: :boom:

**Comment by DavidKang** [comment]
@ChrisBr, could you re-review it again, please. I added the missing test that you told me yesterday.
Thanks :smile_cat:
